### PR TITLE
ci/gha: add i386 unit test

### PR DESCRIPTION
**Pull request by kolyshkin**
Run unit tests on i386. This might help to uncover some issues in the code related to 32-bit arches.

### CODE REVIEW

**Comment by kolyshkin** [comment]
`TestNsenterValidPaths` timeouts:

```
=== RUN   TestNsenterValidPaths
panic: test timed out after 3m0s

goroutine 20 [running]:
testing.(*M).startAlarm.func1()
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1618 +0xcb
created by time.goFunc
	/opt/hostedtoolcache/go/1.15.6/x64/src/time/sleep.go:167 +0x37

goroutine 1 [chan receive]:
testing.(*T).Run(0xa493340, 0x83e0855, 0x15, 0x83eec00, 0x402)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1169 +0x230
testing.runTests.func1(0xa493260)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1439 +0x5a
testing.tRunner(0xa493260, 0xa4b8eec)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0xb7
testing.runTests(0xa49c450, 0x86246a0, 0x4, 0x4, 0x16c72f37, 0xbffc3e11, 0xe924f96a, 0x29, 0x8628840, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1437 +0x25a
testing.(*M).Run(0xa4c6140, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1345 +0x178
main.main()
	_testmain.go:49 +0x104

goroutine 19 [syscall]:
syscall.Syscall(0x3, 0x8, 0xa4f2400, 0x200, 0xa429500, 0xc, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/syscall/asm_linux_386.s:19 +0x5
syscall.read(0x8, 0xa4f2400, 0x200, 0x200, 0x1052a4f, 0x800000, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/syscall/zsyscall_linux_386.go:686 +0x49
syscall.Read(...)
	/opt/hostedtoolcache/go/1.15.6/x64/src/syscall/syscall_unix.go:187
internal/poll.(*FD).Read.func1(0x7ff, 0x2, 0xa491548)
	/opt/hostedtoolcache/go/1.15.6/x64/src/internal/poll/fd_unix.go:155 +0x3a
internal/poll.ignoringEINTR(0xa4b9e20, 0x72, 0xa4f2401, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/internal/poll/fd_unix.go:567 +0x1d
internal/poll.(*FD).Read(0xa491540, 0xa4f2400, 0x200, 0x200, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/internal/poll/fd_unix.go:155 +0x11f
os.(*File).read(...)
	/opt/hostedtoolcache/go/1.15.6/x64/src/os/file_posix.go:31
os.(*File).Read(0xa482cc0, 0xa4f2400, 0x200, 0x200, 0x83a20a0, 0xa49c490, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/os/file.go:116 +0x58
encoding/json.(*Decoder).refill(0xa4d4480, 0xf7fb8ce8, 0x4)
	/opt/hostedtoolcache/go/1.15.6/x64/src/encoding/json/stream.go:165 +0xe8
encoding/json.(*Decoder).readValue(0xa4d4480, 0x0, 0x0, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/encoding/json/stream.go:140 +0x1d8
encoding/json.(*Decoder).Decode(0xa4d4480, 0x8353180, 0xa482d10, 0xa487fa0, 0x0)
	/opt/hostedtoolcache/go/1.15.6/x64/src/encoding/json/stream.go:63 +0x65
github.com/opencontainers/runc/libcontainer/nsenter.TestNsenterValidPaths(0xa493340)
	/home/runner/work/runc/runc/libcontainer/nsenter/nsenter_test.go:71 +0x650
testing.tRunner(0xa493340, 0x83eec00)
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1123 +0xb7
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.15.6/x64/src/testing/testing.go:1168 +0x211
FAIL	github.com/opencontainers/runc/libcontainer/nsenter	180.013s
```

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -69,6 +69,10 @@ jobs:
       if: matrix.rootless != 'rootless'
       run: sudo -E PATH="$PATH" -- make localunittest
 
+    - name: unit test (i386)
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:72]: I think we should clarify the purpose of this in the YAML:
```
# We need to continue support for 32-bit ARM. However, we do not have 32-bit ARM CI, so we use i386 for testing 32bit stuff.
# We are not interested in providing official support for i386.
```

Comment (kolyshkin) [reply .github/workflows/test.yml:72]: Unfortunately everything that is nsenter / runc init related is not working. I tried to debug this but got nowhere so I'm putting this on hold for now. :(

Yet your point is valid; updated.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -69,6 +69,13 @@ jobs:
       if: matrix.rootless != 'rootless'
       run: sudo -E PATH="$PATH" -- make localunittest
 
+    # We need to continue support for 32-bit ARM.
+    # However, we do not have 32-bit ARM CI, so we use i386 for testing 32bit stuff.
+    # We are not interested in providing official support for i386.
+    - name: unit test (i386)
+      if: matrix.rootless != 'rootless'
+      run: sudo -E PATH="$PATH" -- make GOARCH=386 localunittest
```
Comment (zhsj) [inline .github/workflows/test.yml:77]: the nsenter test doesn't work in cross build. In cross build, cgo is disabled, so `nsenter_unsupported.go` is used.

However it should work if running the test in docker, with `--platform=386` args.

```
$ docker run --privileged -it --rm --platform 386 -v `pwd`:/work golang:1.15-buster bash
root@3084de9b4c8f:/go# cd /work/libcontainer/nsenter
root@3084de9b4c8f:/work/libcontainer/nsenter# go test -v .
=== RUN   TestNsenterValidPaths
--- PASS: TestNsenterValidPaths (0.00s)
=== RUN   TestNsenterInvalidPaths
--- PASS: TestNsenterInvalidPaths (0.00s)
=== RUN   TestNsenterIncorrectPathType
--- PASS: TestNsenterIncorrectPathType (0.00s)
=== RUN   TestNsenterChildLogging
--- PASS: TestNsenterChildLogging (0.00s)
PASS
ok      github.com/opencontainers/runc/libcontainer/nsenter     0.012s
```

Another way it seems cgo can be manually enabled with `CGO_ENABLED=1` env, but need to figure out how to setup gcc cross build env.

Comment (kolyshkin) [reply .github/workflows/test.yml]: Thanks a lot @zhsj, I made it running!

**Comment by kolyshkin** [comment]
It's up and running now, and uncovered a set of issues: https://github.com/opencontainers/runc/issues/2783

**Comment by kolyshkin** [comment]
OK, it's working now.

**Comment by kolyshkin** [comment]
Depend on #2741, waiting for it to be merged.

**Review by AkihiroSuda — APPROVED** [review]

**Review by mrunalp — APPROVED** [review]

**Comment by kolyshkin** [comment]
Failure in centos is a flake (https://github.com/opencontainers/runc/issues/2760); restarted.

**Pull request merged**